### PR TITLE
Add PyArrow 0.15 support back, and test PyArrow 0.16 in CI

### LINKED ISSUE #1109: Re-enable PyArrow 0.15 support back
We should enable https://github.com/databricks/koalas/pull/1108 back once we have enough documentation and error messages.

### PR DESCRIPTION
**Pull request by HyukjinKwon**
This PR adds PyArrow 0.15 support back with a better documentation and error message.
Error message will look as below:

Warning when pyarrow>=0.15 and pyspark<3.0 but `ARROW_PRE_0_15_IPC_FORMAT` not set.

```
WARNING:root:'ARROW_PRE_0_15_IPC_FORMAT' environment variable was not set. It is required 
to set this environment variable to '1' if you use pyarrow>=0.15 and pyspark<3.0. Koalas will 
set it for you but it does not work if there is a Spark context already launched.
```

Exceptions when `ARROW_PRE_0_15_IPC_FORMAT` is set in valid cases:

```
RuntimeError: Please explicitly unset 'ARROW_PRE_0_15_IPC_FORMAT' environment variable 
in both driver and executor sides. Check your spark.executorEnv.*, spark.yarn.appMasterEnv.*, 
spark.mesos.driverEnv.* and spark.kubernetes.driverEnv.* configurations. It is required to set 
this environment variable only when you use pyarrow>=0.15 and pyspark<3.0.
```

```
RuntimeError: Please explicitly unset 'ARROW_PRE_0_15_IPC_FORMAT' environment variable in 
both driver and executor sides. It is required to set this environment variable only when you use 
pyarrow>=0.15 and pyspark<3.0.
```

Resolves #1109

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -54,6 +54,9 @@ dbutils.library.restartPython()
 
 Note that Koalas requires Databricks Runtime 5.x or above. In the future, we will package Koalas out-of-the-box in both the regular Databricks Runtime and Databricks Runtime for Machine Learning.
 
+Lastly, note that if your PyArrow version is 0.15+, it is best for you to set `ARROW_PRE_0_15_IPC_FORMAT` environment variable to `1` manually.
```
Comment (ueshin) [inline README.md:57]: Maybe we should mention PySpark version as well. It's only needed when PyArrow>=0.15 and PySpark<3.

File: databricks/koalas/__init__.py [hunk 2]
```diff
@@ -48,9 +48,26 @@ def assert_pyspark_version():
...
-    # This is required to support PyArrow 0.15 in PySpark versions lower than 3.0.
-    # See SPARK-29367.
-    os.environ["ARROW_PRE_0_15_IPC_FORMAT"] = "1"
+    if "ARROW_PRE_0_15_IPC_FORMAT" not in os.environ:
+        import logging
+
+        logging.warning(
+            "'ARROW_PRE_0_15_IPC_FORMAT' environment variable was not set. It is required to "
+            "set this environment variable to '1' in both driver and executor sides if you use "
+            "pyarrow>=0.15 and pyspark<=3.0. "
```
Comment (ueshin) [inline databricks/koalas/__init__.py:57]: `pyspark<3.0`

File: databricks/koalas/__init__.py [hunk 3]
```diff
@@ -48,9 +48,26 @@ def assert_pyspark_version():
...
+            "launched."
+        )
+        # This is required to support PyArrow 0.15 in PySpark versions lower than 3.0.
+        # See SPARK-29367.
+        os.environ["ARROW_PRE_0_15_IPC_FORMAT"] = "1"
+elif "ARROW_PRE_0_15_IPC_FORMAT" in os.environ:
+    raise RuntimeError(
+        "Please explicitly unset 'ARROW_PRE_0_15_IPC_FORMAT' environment variable in both "
+        "driver and executor sides. It is required to set this environment variable only "
+        "when you use pyarrow>=0.15 and pyspark<=3.0."
```
Comment (ueshin) [inline databricks/koalas/__init__.py:68]: ditto.

File: databricks/koalas/utils.py [hunk 4]
```diff
@@ -296,20 +297,43 @@ def align_diff_series(func, this_series, *args, how="full"):
...
+                session.conf.get("spark.kubernetes.driverEnv.ARROW_PRE_0_15_IPC_FORMAT", None),
+            ]
+        )
+        if is_legacy_ipc_set:
+            raise RuntimeError(
+                "Please explicitly unset 'ARROW_PRE_0_15_IPC_FORMAT' environment variable in "
+                "both driver and executor sides. Check your spark.executorEnv.*, "
+                "spark.yarn.appMasterEnv.*, spark.mesos.driverEnv.* and "
+                "spark.kubernetes.driverEnv.* configurations. It is required to set this "
+                "environment variable only when you use pyarrow>=0.15 and pyspark<=3.0."
```
Comment (ueshin) [inline databricks/koalas/utils.py:334]: ditto.

File: databricks/koalas/__init__.py [hunk 5]
```diff
@@ -48,9 +48,26 @@ def assert_pyspark_version():
...
+            "'ARROW_PRE_0_15_IPC_FORMAT' environment variable was not set. It is required to "
+            "set this environment variable to '1' in both driver and executor sides if you use "
+            "pyarrow>=0.15 and pyspark<=3.0. "
+            "Koalas will set it for you but it does not work if there is a Spark context already "
+            "launched."
+        )
+        # This is required to support PyArrow 0.15 in PySpark versions lower than 3.0.
+        # See SPARK-29367.
+        os.environ["ARROW_PRE_0_15_IPC_FORMAT"] = "1"
+elif "ARROW_PRE_0_15_IPC_FORMAT" in os.environ:
```
Comment (ueshin) [inline databricks/koalas/__init__.py:64]: The case `pyarrow<0.15` and `pyspark<3.0` falls to this clause.

Comment (ueshin) [reply databricks/koalas/__init__.py:64]: maybe:

```py

if LooseVersion(pyspark.__version__) < LooseVersion("3.0"):
    if (
        LooseVersion(pyarrow.__version__) >= LooseVersion("0.15")
        and "ARROW_PRE_0_15_IPC_FORMAT" not in os.environ
    ):
        import logging
        ...

elif "ARROW_PRE_0_15_IPC_FORMAT" in os.environ:
    raise RuntimeError(
        ...)
```

?

File: databricks/koalas/utils.py [hunk 6]
```diff
@@ -18,6 +18,7 @@
 """
 
 import functools
+import os
```
Comment (ueshin) [inline databricks/koalas/utils.py:21]: nit: not used?

**Comment by HyukjinKwon** [comment]
Thanks for detailed review @ueshin!

File: databricks/koalas/__init__.py (hunk 2 above)
Comment (ueshin) [reply databricks/koalas/__init__.py:58]: This is not addressed yet?

File: databricks/koalas/__init__.py (hunk 3 above)
Comment (ueshin) [reply databricks/koalas/__init__.py:69]: ditto.

File: databricks/koalas/__init__.py (hunk 2 above)
Comment (HyukjinKwon) [reply databricks/koalas/__init__.py]: Oops sorry, I just addressed.

**Comment by ueshin** [comment]
Thanks! merging.

**Pull request merged**